feat(recovery): account recovery phase 1.5 setup screens

vbudhram opened the pull request:
Extracted from https://github.com/mozilla/fxa-content-server/pull/6203, so that it is more manageable to review. This PR only contains account recovery setup screens and unit tests so you can create and revoke recovery keys for a user. To view the screens append `?showAccountRecovery` to the url.

In a follow up issue, I'll add metrics and experiment logic, ref https://github.com/mozilla/fxa-content-server/issues/6394.

vbudhram commented on app/scripts/templates/settings/account_recovery/account_recovery.mustache:
```diff
@@ -0,0 +1,54 @@
+
+    <div class="settings-unit-details">
+        <p>
+            {{#unsafeTranslate}}Your Firefox Account password is used to secure all of the information you’ve saved to Firefox. If you need to reset your password, you’ll need a recovery key to regain access to this information. <a>Learn more</a>{{/unsafeTranslate}}
```
SUMO article is being drafted [here](https://docs.google.com/document/d/1WgEE6CzTrYj6MEr_4xRRnH90jGqvggGxo35ROwfLZbg/edit?usp=sharing), once ready this link can be updated.

vbudhram commented on app/scripts/views/settings/account_recovery/recovery_key.js:
```diff
@@ -0,0 +1,79 @@
+    'click .print-key': '_printKey'
+  },
+
+  _getFormatedRecoveryKeyFilename() {
```
I thought about extracting print and download into their own mixin but the code churn was a bit much. We should consider this as a followup issue.

vbudhram commented on package.json:
```diff
@@ -53,7 +53,7 @@
     "fxa-crypto-relier": "2.3.0",
     "fxa-geodb": "1.0.2",
-    "fxa-js-client": "1.0.4",
+    "fxa-js-client": "git@github.com:mozilla/fxa-js-client.git#master",
```
Flagging myself to make a new tag before this get merged.

vbudhram commented:
@mozilla/fxa-devs I think this is ready for review!

shane-tomlinson commented on app/scripts/templates/settings/account_recovery/confirm_password.mustache:
```diff
@@ -0,0 +1,30 @@
+            <h2 class="email">{{email}}</h2>
+
+            <div class="input-row password-row">
+                <input type="password" class="password tooltip-below" id="password" placeholder="{{#t}}Enter Password{{/t}}" required pattern=".{8,}" autofocus/>
```
The `autofocus` isn't kicking in here, not sure why. I have to manually click into the field.

shane-tomlinson commented on app/scripts/lib/fxa-client.js:
```diff
@@ -1034,7 +1035,57 @@ define(function (require, exports, module) {
+        .then((res) => client.accountKeys(res.keyFetchToken, res.unwrapBKey))
+        .then((result) => {
+          keys = result;
+          return RecoveryKey.generateRecoveryKey(length)
```
I'm unable to create a recovery key bundle using the memory DB.

On the auth-server, a 500 is returned and the logs show:

<img width="997" alt="screen shot 2018-07-30 at 12 10 13" src="https://user-images.githubusercontent.com/848085/43394378-76a015c2-93f2-11e8-8fcf-c96eab685412.png">

On the auth-db, I see:

<img width="989" alt="screen shot 2018-07-30 at 12 15 46" src="https://user-images.githubusercontent.com/848085/43394389-7de9a85c-93f2-11e8-9a3e-43ef5966cc2a.png">

vbudhram replied:
Are you running the latest auth-server master? This doesn't fail for me locally with either mem or mysql.

shane-tomlinson replied:
Yup, I did a `_scripts/update_all.sh` just in case.

vbudhram replied:
In your local db server, is this [line](https://github.com/mozilla/fxa-auth-db-mysql/blob/29b9b4bc65784659ff103d77a390dc043e41992c/db-server/index.js#L242) set to `/account/:id/recoveryKey` or `/account/:id/recoveryKeys`?  I made this update a little while back to make the route singular and suspect this is why it is not finding it.

shane-tomlinson commented on app/scripts/templates/settings/account_recovery/recovery_key.mustache:
```diff
@@ -0,0 +1,33 @@
+        <h2>{{#t}}Recovery key{{/t}}</h2>
+
+        <p class="description">
+            <span class="note">{{#t}}NOTE:{{/t}}</span>
```
The same comment about `Important:` from above applies here.

shane-tomlinson commented on app/scripts/templates/settings/account_recovery/recovery_key.mustache:
```diff
@@ -0,0 +1,33 @@
+            </div>
+
+            <div class="buttons">
+                <button class="download-key"><img class="recovery-icon-option" src="/images/recovery_key_download.svg">{{#t}}Download{{/t}}</button>
```
Neither of these buttons have styles, is that expected?

vbudhram replied:
The styles get putted in [here](https://github.com/mozilla/fxa-content-server/blob/823d5983732db81ee669a3ea8eaa09fc125f3620/app/styles/modules/_settings-account-recovery.scss#L113). I could give them an explicit class though.

shane-tomlinson replied:
>> Neither of these buttons have styles, is that expected?

> I could give them an explicit class though.

Poor choice of wording on my part. I meant the buttons are both white with a grey outline, is that the style you are going for?

vbudhram replied:
Ah, yea that is the style I am going for.

shane-tomlinson commented on app/scripts/templates/settings/account_recovery/recovery_key.mustache:
```diff
@@ -0,0 +1,33 @@
+        </div>
+
+        <div class="button-row">
+            <button class="secondary-button done-link">{{#t}}Done, Recovery Key Set{{/t}}</button>
```
What might be a nice touch is to convert this from a secondary button to a primary button once the user either downloads or prints. The secondary button seems a little strange to me though.

vbudhram replied:
That might be the intended flow, I'll follow up in https://github.com/mozilla/fxa-content-server/issues/6394.

shane-tomlinson commented on app/scripts/templates/settings/account_recovery/confirm_revoke.mustache:
```diff
@@ -0,0 +1,14 @@
+        <form novalidate>
+
+            <h3>{{#t}}Are you sure you want to revoke?{{/t}}</h3>
+            <p class="revoke-description">{{#t}}Revoking your account recovery key will not allow it to be used to recover your data in the event you reset your password.{{/t}}</p>
```
I find the `will not allow it to be used` confusing and the "Are you sure you want to revoke?" makes "Revoking your account recovery key" seem redundant.

Could it be shorter/simpler?

> You will be unable to recovery your data in the event you reset your password.


Or...

> You will be unable to use your recovery key to access your data in the event you reset your password.

shane-tomlinson commented on app/scripts/templates/settings/account_recovery/recovery_key.mustache:
```diff
@@ -0,0 +1,33 @@
+
+        <p class="description">
+            <span class="note">{{#t}}NOTE:{{/t}}</span>
+            {{#t}}Keep your recovery key in a safe place where you can access it later. You will not be able to access this page again. You can download or print your key or email it to someone you trust.{{/t}}
```
For the last sentence, I'd ditch the `You can` to make it more imperative that the user do something with the key.

shane-tomlinson commented on app/scripts/views/settings/account_recovery/confirm_revoke.js:
```diff
@@ -0,0 +1,38 @@
+  className: 'account-recovery-confirm-revoke',
+  viewName: 'settings.account-recovery.confirm-revoke',
+
+  events: {
```
You can access this page w/o having set a recovery key just by browing to `/settings/account_recovery/confirm_revoke`. Maybe in a `beforeRender` or `afterRender`, ensure the user *has* a recovery key or else send them to the setup page.

vbudhram replied:
Added!

shane-tomlinson commented on app/scripts/views/settings/account_recovery/confirm_revoke.js:
```diff
@@ -0,0 +1,38 @@
+
+  submit() {
+    const account = this.getSignedInAccount();
+    return account.deleteRecoveryKey()
```
This does not display network errors. I tried to delete the account recovery key on an account where it wasn't set, I see a 500 error in the console but no error in the UI.

vbudhram replied:
I wasn't able to replicate this locally with those steps. The delete request 500?

vbudhram replied:
This might be related to https://github.com/mozilla/fxa-content-server/pull/6392#discussion_r206270298, this route was also updated to be singular.

shane-tomlinson replied:
You are right, the shrinkwrap file for the auth-server needs to be updated with the newest db-server.

shane-tomlinson commented on app/scripts/views/settings/account_recovery/recovery_key.js:
```diff
@@ -0,0 +1,79 @@
+  className: 'account-recovery-key',
+  viewName: 'settings.account-recovery.recovery-key',
+
+  events: {
```
Like above, I can access this page w/o having set a recovery key just by browing to `/settings/account_recovery/recovery_key`. Maybe in a `beforeRender` or `afterRender`, ensure the user *has* a recovery key or else send them to the setup page.

vbudhram replied:
Added!

shane-tomlinson commented on app/scripts/views/settings/account_recovery/confirm_password.js:
```diff
@@ -0,0 +1,55 @@
+        this.showRecoveryKeyView = true;
+        this.navigate('settings/account_recovery/recovery_key', result);
+      })
+      .catch((err) => this.showValidationError(this.$(PASSWORD_SELECTOR), err));
```
This will show *any* error as a validation error. That seems wrong to me. An invalid password error should be a validation error, others seem like they should be a general panel error.

vbudhram commented on app/scripts/views/mixins/modal-settings-panel-mixin.js:
```diff
@@ -37,16 +37,42 @@ define(function (require, exports, module) {
-    onModalCancel () {
-      if (this.currentPage === 'settings/clients/disconnect') {
+    onModalCancel() {
+      switch (this.currentPage) {
```
Yea it is strange. I actually think for this case we should have a different type of mixin, once that can handle wizard type modals. I noticed that `onModalCancel` gets called whenever the modal disappears, which could be transition to next modal or a cancel.

vbudhram commented on app/tests/spec/views/settings/account_recovery/account_recovery.js:
```diff
@@ -0,0 +1,127 @@
+const View = require('views/settings/account_recovery/account_recovery');
+
+describe('views/settings/account_recovery/account_recovery', () => {
+  let account, broker, email, metrics, notifier, hasRecoveryKey, relier, sentryMetrics, user, view, showAccountRecovery;
```
Should we? I don't have a big preference either way, just thought it looked strange to have so many variable initialization lines.

vbudhram commented on app/scripts/views/settings/account_recovery/recovery_key.js:
```diff
@@ -0,0 +1,79 @@
+    const template = `
+      <a id="recovery-key-download-link" href="${href}" download="${this._getFormatedRecoveryKeyFilename()}"></a>
+    `;
+    $(template).appendTo('#account-recovery-key');
```
I tried this with `this.$` and it complained about not being to be able to render template. 

```
Error: Syntax error, unrecognized expression: 
      <a id="recovery-key-download-link" href="blob:http://127.0.0.1:3030/c3b5e51c-de8a-4f42-8ed6-23120a806d5a" download="a@a.com Firefox Recovery Key.txt"></a>
```

Looking through the code, when using the `appendTo` jquery function, we use `$` directly, so there is probably a reason why it is this way. Maybe similar to https://github.com/mozilla/fxa-content-server/pull/5992#discussion_r178301603.

vbudhram commented on app/scripts/templates/settings/account_recovery/confirm_password.mustache:
```diff
@@ -0,0 +1,30 @@
+                <button class="primary-button generate-key-link">{{#t}}Continue ->{{/t}}</button>
+            </div>
+            <div class="link-options">
+                <a class="cancel-link">{{#t}}Cancel{{/t}}</a>
```
Updated this to be a button.

vbudhram commented:
@shane-tomlinson Thanks for review! I think I have address all of the comments except the ones I think might be related to https://github.com/mozilla/fxa-content-server/pull/6392#discussion_r206270298.

vbudhram commented on app/scripts/views/settings/account_recovery/account_recovery.js:
```diff
@@ -0,0 +1,84 @@
+  className: 'account-recovery',
+  viewName: 'settings.account-recovery',
+
+  events: {
```
Turns out this didn't need to be a form, changed to baseview.

shane-tomlinson commented on app/scripts/templates/settings/account_recovery/confirm_password.mustache:
```diff
@@ -0,0 +1,27 @@
+
+        <form novalidate>
+            <p class="description">
+                {{#t}}Enter your password to view your recovery key.{{/t}}
```
Some thoughts about this screen.

1. The "Enter your password to view your recovery key" looks out of place up there and is almost totally detached from the input element.
2. The email looks exceptionally heavy.
3. The important text kinds gets lost. It might be worthwhile to experiment [with the Photon styles for warnings](https://design.firefox.com/photon/patterns/warnings.html) to see if the yellow color and warning icon work well here.

<img width="402" alt="screen shot 2018-07-31 at 15 45 59" src="https://user-images.githubusercontent.com/848085/43469586-69646a2e-94de-11e8-9cca-550615d24197.png">

Here's a quick mock changing the order, making the email an h2, removing the current "re-enter your password to...." text, and adding the yellow color (I'm not being ambitious enough to try the icon):

<img width="463" alt="screen shot 2018-07-31 at 16 33 16" src="https://user-images.githubusercontent.com/848085/43470055-8e90078a-94df-11e8-8d94-930bd9dd527c.png">

vbudhram replied:
@shane-tomlinson This differs a lot from originally proposed UX in Phase 1.5, I can try to get some additional info from mliang. In meantime, I can flag it in https://github.com/mozilla/fxa-content-server/issues/6394 for follow up.

youwenliang replied:
@vbudhram I don't have the "important text" part in my UX spec, do you know where that came from?

I agree that the visual hierarchy could be better (with the important message) but I think the photon style yellow background is not suitable here since the message here is not "warning" to me and it might unnecessarily alert the users.

If we do need that important message, I can provide an updated version of this screen.

vbudhram replied:
@youwenliang I can't seem to find the version of UX that contained this. Do you have a final UX doc with copy changes? Best I have is [this](https://mozilla.invisionapp.com/share/RMMYSS8C4Z3#/screens/307726824).

In the meantime, I can remove the important block.

shane-tomlinson commented on app/scripts/templates/settings/account_recovery/confirm_revoke.mustache:
```diff
@@ -0,0 +1,15 @@
+
+            <div class="button-row">
+                <button action="submit" class="settings-button warning-button revoke">{{#t}}Revoke{{/t}}</button>
+                <button class="settings-button secondary-button cancel-link">{{#t}}Cancel{{/t}}</button>
```
The link here looks very out of place. Other panels use a grey secondary button, e.g., setting an avatar via the camera:

<img width="411" alt="screen shot 2018-07-31 at 16 36 16" src="https://user-images.githubusercontent.com/848085/43470164-dc26abca-94df-11e8-8104-90cf24a9b66a.png">
<img width="426" alt="screen shot 2018-07-31 at 16 31 13" src="https://user-images.githubusercontent.com/848085/43470168-de4483a0-94df-11e8-9f6d-195c7df21441.png">

Delete account uses a grey secondary button too:

<img width="503" alt="screen shot 2018-07-31 at 15 47 06" src="https://user-images.githubusercontent.com/848085/43470211-f10510fe-94df-11e8-985c-f31a59d5ded2.png">

shane-tomlinson commented on app/scripts/templates/settings/account_recovery/recovery_key.mustache:
```diff
@@ -0,0 +1,32 @@
+        </p>
+
+        <div class="recovery-key-container">
+            <div class="recovery-key">{{recoveryKey}}</div>
```
The recovery key could be split into sections of 4 like a credit card to make it easier to write:

<img width="339" alt="screen shot 2018-07-31 at 16 38 08" src="https://user-images.githubusercontent.com/848085/43470286-1dde0d88-94e0-11e8-8895-6b8580ac05e9.png">

vs

<img width="362" alt="screen shot 2018-07-31 at 16 39 07" src="https://user-images.githubusercontent.com/848085/43470351-44e21168-94e0-11e8-9ea1-7929ea9163be.png">

shane-tomlinson replied:
This change looks great, thanks!

shane-tomlinson commented on app/scripts/templates/settings/account_recovery/recovery_key.mustache:
```diff
@@ -0,0 +1,32 @@
+            </div>
+
+            <div class="buttons">
+                <button class="download-key"><img alt="Download button image" class="recovery-icon-option" src="/images/recovery_key_download.svg">{{#t}}Download{{/t}}</button>
```
The alt text needs the `{{#t}}` wrapper here and on the next button.

shane-tomlinson commented on app/scripts/templates/settings/account_recovery/recovery_key_print.mustache:
```diff
@@ -0,0 +1,7 @@
+<head><title>Recovery Key</title>
+</head>
+<body>
+    <p>{{recoveryKey}}</p>
```
Since the user has to manually enter the recovery code, I'd split this into sections of 4 as well.

shane-tomlinson commented on app/scripts/views/settings/account_recovery/confirm_password.js:
```diff
@@ -0,0 +1,62 @@
+  className: 'account-recovery-confirm-password',
+  viewName: 'settings.account-recovery.confirm-password',
+
+  events: {
```
The comment about [extending FormView's events](https://github.com/mozilla/fxa-content-server/pull/6392/files#r206087360) applies here too.

vbudhram replied:
@shane-tomlinson Not sure I follow what you mean, should it be extended like [this](https://github.com/mozilla/fxa-content-server/blob/3573ecf044f1774e4bdfae736f126f1ecad1a8f8/app/scripts/views/force_auth.js#L157)?

shane-tomlinson commented on app/scripts/views/settings/account_recovery/confirm_revoke.js:
```diff
@@ -0,0 +1,48 @@
+  className: 'account-recovery-confirm-revoke',
+  viewName: 'settings.account-recovery.confirm-revoke',
+
+  events: {
```
Same comment about extending FormView's events applies here too.

shane-tomlinson commented on app/tests/spec/views/settings/account_recovery/recovery_key.js:
```diff
@@ -0,0 +1,166 @@
+    broker = new Broker();
+    email = TestHelpers.createEmail();
+    notifier = new Notifier();
+    sentryMetrics = new SentryMetrics();
```
is sentryMetrics actually used anywhere? If not, you could just pass in a mock? (Applies to the other tests too)

shane-tomlinson commented on app/styles/modules/_settings-account-recovery.scss:
```diff
@@ -0,0 +1,165 @@
+        }
+      }
+
+      .settings-button-group {
```
The `.details` and `.settings-button-group` blocks are duplicated 4x in the scss now, they should be extracted to reduce duplication.

vbudhram replied:
I combined the `.details` for emails, account recovery and totp views. However, for `.settings-button-group` I can only combine account recovery and totp since they are similar panels in terms of button layout. The emails panel has multiple buttons on the row and would have caused the scss to be kinda wonky.

shane-tomlinson commented on app/scripts/templates/settings/account_recovery/confirm_password.mustache:
```diff
@@ -0,0 +1,24 @@
+                {{#t}}Enter your password to view your recovery key.{{/t}}
+            </p>
+
+            <h2 class="email">{{email}}</h2>
```
If the styling for this element is retained, can you change this from an h2 since this isn't really a header?

shane-tomlinson commented on app/scripts/views/settings/account_recovery/account_recovery.js:
```diff
@@ -0,0 +1,85 @@
+  setInitialContext(context) {
+    const hasRecoveryKey = this.model.get('hasRecoveryKey');
+    context.set({
+      escapedLearnMoreLinkAttributes: _.escape('target=_blank href=' + ACCOUNT_RECOVERY_SUPPORT_URL),
```
_.escape is usually to HTML escape a block of text and will break if the `ACCOUNT_RECOVERY_SUPPORT_URL` contains an `&` since that'll be converted to `&amp;`. It's better to call `encodeURI` on the URLs, or even use [xss.js->href](https://github.com/mozilla/fxa-content-server/blob/master/app/scripts/lib/xss.js#L15)

```js
escapedLearnMoreLinkAttributes: `target="_blank" href="${encodeURI(ACCOUNT_RECOVERY_SUPPORT_URL)}"`
```

shane-tomlinson commented on app/scripts/views/settings/account_recovery/recovery_key.js:
```diff
@@ -0,0 +1,98 @@
+  _formatRecoveryKey(key) {
+    if (key) {
+      // Insert spaces every 4 characters and remove trailing space
+      return key.replace(/(\w{4})/g, '$1 ').replace(/(^\s+|\s+$)/, '');
```
This is pretty awesome, I've not seen this way of chunking up a string before.

shane-tomlinson commented on app/scripts/views/settings/account_recovery/recovery_key.js:
```diff
@@ -0,0 +1,98 @@
+
+  _getFormatedRecoveryKeyFilename() {
+    const account = this.getSignedInAccount();
+    let formattedFilename = account.get('email') + ' ' + t('Firefox Recovery Key');
```
Can filenames contain emoji characters? Email addresses can if created programatically, though we do not yet allow that on the front end.

vbudhram replied:
Tested on OSX and filenames can have emoji characters. I think this might be ok for now.

shane-tomlinson commented on app/scripts/views/settings/account_recovery/account_recovery.js:
```diff
@@ -0,0 +1,85 @@
+  }, CODE_REFRESH_SELECTOR, CODE_REFRESH_DELAY_MS),
+
+  cancel() {
+    return this.render()
```
How come the render right before navigating?

shane-tomlinson commented on app/scripts/templates/settings/account_recovery/recovery_key.mustache:
```diff
@@ -0,0 +1,32 @@
+            </div>
+
+            <div class="buttons">
+                <button class="download-key"><img alt="{{#t}}Download button image{{/t}}" class="recovery-icon-option" src="/images/recovery_key_download.svg">{{#t}}Download{{/t}}</button>
```
I noticed that neither of these buttons have visible hover states, is that intentional? The cursor changes, but no color changes.

![button-hover](https://user-images.githubusercontent.com/848085/43536109-ab17429a-95b3-11e8-94e2-3ee9e898fb52.gif)

vbudhram replied:
Hmm not sure, but I'll flag for UX follow up in https://github.com/mozilla/fxa-content-server/issues/6394.

shane-tomlinson commented:
One thing I forgot to say.. Functional tests... Those are definitely needed.

vbudhram commented:
@shane-tomlinson Thanks for review! I made updates from latest round and added some in the follow up issue. I also added a bare bones functional test.

SorinaFlorean commented:
Verified on staging server - Train 118.
